project/src/main/java/com/google/sps/utilities/SpannerTasks.java

```diff
@@ -600,4 +606,84 @@ public Void run(TransactionContext transaction) throws Exception {
     }
     return results;
   }
+ 
+  /**
+   * Add event results to persistent storage index by inserting the keyword into the keywords table
+   * if it dose not exist and adding a corresponding row in the results table.
+   *
+   * @param results the event results to add to the index
+   */
+  public static void addResultsToPersistentStorageIndex(List<EventResult> results) {
+    for (EventResult result : results) {
+      SpannerClient.getDatabaseClient()
+          .readWriteTransaction()
+          .run(
+              new TransactionCallable<Void>() {
+                @Override
+                public Void run(TransactionContext transaction) throws Exception {
+                  String eventId = result.getEventId();
+                  String keyword = result.getKeyword();
+                  float ranking = result.getRanking();
+                  Struct row =
+                      transaction.readRow(
+                          KEYWORDS_TABLE, Key.of(keyword), Collections.singleton(NAME));
+                  String keywordId = row == null ? UUID.randomUUID().toString() : row.getString(0);
+                  transaction.buffer(
+                      Mutation.newInsertBuilder(KEYWORDS_TABLE)
+                          .set(KEYWORD_ID)
+                          .to(keywordId)
+                          .set(NAME)
+                          .to(keyword)
+                          .build());
+                  transaction.buffer(
+                      Mutation.newInsertBuilder(RESULTS_TABLE)
+                          .set(KEYWORD_ID)
+                          .to(keywordId)
+                          .set(EVENT_ID)
+                          .to(eventId)
+                          .set(RANKING)
+                          .to(ranking)
+                          .build());
+                  return null;
+                }
+            });
+    }
+  }
+
+  /**
+   * Returns events by keyword in descending order of ranking.
+   *
+   * @param keyword keyword to fetch event results for
+   * @return list of events in descending order of ranking
+   */
+  public static List<Event> getEventResultsByKeywordWithDescendingRanking(String keyword) {
+    List<Event> results = new ArrayList<Event>();
+    Statement statement =
+        Statement.of(
+            String.format(
+                "SELECT Events.EventID, Events.Name, Events.Description, Events.Labels,"
+                    + " Events.Location, Events.Date, Events.Time, Events.Host FROM Results INNER"
+                    + " JOIN Keywords ON Results.KeywordID = Keywords.KeywordID INNER JOIN Events"
+                    + " ON Events.EventID = Results.EventID WHERE Keywords.Name=\"%s\" ORDER BY"
+                    + " Results.Ranking DESC LIMIT 20;",
```

Comment:
is there a reason why there's a limit to # of events returned?

Reply:
I am planning on showing a maximum of 20 results for a search.